### PR TITLE
[ES-7757] Convert package to typescript

### PR DESCRIPTION
**Pull request by Sisyphus25**
- Convert lib to ts
- Fix test and demo build
- Allow build and import the tracker as single js script
- Update package and remove some copy paste stuff

### CODE REVIEW

File: fixup [hunk 1]
```diff
@@ -0,0 +1,11 @@
+cat >cjs/package.json <<!EOF
```
Comment (wkurniawan07) [inline fixup:1]: Unnecessary file?

File: index.html [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+  </head>
+
+  <body onload="myfunction()">
+    <!--[if lt IE 10]>
+      <p class="browsehappy">
+        You are using an <strong>outdated</strong> browser. Please
+        <a href="http://browsehappy.com/">upgrade your browser</a> to improve
+        your experience.
+      </p>
+    <![endif]-->
```
Comment (wkurniawan07) [inline index.html:31]: I don't think we need any of these IE stuff anymore. IE is formally EOL.

File: package.json [hunk 3]
```diff
@@ -1,27 +1,30 @@
 {
   "name": "visenze-tracking-javascript",
-  "version": "1.2.1",
+  "version": "2.0.0",
```
Comment (wkurniawan07) [inline package.json:3]: Let's make it rc version first, so we are certain all the integrations work.
```suggestion
  "version": "2.0.0-rc.0",
```

File: tsconfig.base.json [hunk 4]
```diff
@@ -0,0 +1,15 @@
+{
+  "include": ["src/**/*", "types/**/*"],
+  "compilerOptions": {
+    "moduleResolution": "node",
+    "declaration": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "resolveJsonModule": true,
+    "strict": true,
+    "noImplicitAny": false,
```
Comment (wkurniawan07) [inline tsconfig.base.json:10]: Any reason for suppressing this?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Mostly fine, just some changes needed

File: fixup (hunk 1 above)
Comment (Sisyphus25) [reply fixup:1]: this is necessary since esm and cjs have different import entries. You can check the package.json file.

Comment (wkurniawan07) [reply fixup:1]: I believe there's a better way to achieve this, but let's keep this for now

**Review by wkurniawan07 — APPROVED** [review]
LGTM for now, let's try to publish and use in search SDK

File: .eslintignore [hunk 5]
```diff
@@ -0,0 +1,7 @@
+test
+lib
+babel.config.json
+webpack.dev.js
+webpack.prod.js
+csj
```
Comment (wkurniawan07) [inline .eslintignore:6]: Typo; should be cjs?

File: tsconfig.base.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
+  "include": ["src/**/*", "types/**/*"],
+  "compilerOptions": {
+    "moduleResolution": "node",
+    "declaration": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "resolveJsonModule": true,
+    "strict": true,
+    "noImplicitAny": true,
```
Comment (wkurniawan07) [inline tsconfig.base.json:10]: Then this is not needed at all, since `"strict": true` already include this

File: .eslintrc.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    'eslint:recommended',
+    'plugin:@typescript-eslint/recommended',
+    'plugin:@typescript-eslint/recommended-requiring-type-checking',
+  ],
+  parserOptions: {
+    project: './tsconfig.json',
+  },
+  ignorePatterns: ['dist/**/*'],
+  rules: {
+    '@typescript-eslint/no-explicit-any': 'error',
```
Comment (wkurniawan07) [inline .eslintrc.js:18]: Quite surprised to see this achievable. I guess the functions are simple enough and we're lucky to depend on well-maintained libraries.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: .eslintrc.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    '@typescript-eslint/no-unsafe-argument': 'off',
+    '@typescript-eslint/no-unsafe-member-access': 'off',
+    '@typescript-eslint/no-unsafe-assignment': 'off',
+    '@typescript-eslint/no-unsafe-call': 'off',
+    '@typescript-eslint/no-floating-promises': 'off',
```
Comment (wkurniawan07) [inline .eslintrc.js:23]: In addition, none of these suppressed rules should be necessary, now that `@types/node-fetch` exist

Comment (Sisyphus25) [reply .eslintrc.js]: @wkurniawan07 I checked again and it seems fine to remove all the suppressed rules now.

**Review by srfication — APPROVED** [review]

**Review by wkurniawan07 — APPROVED** [review]

File: tsconfig.base.json [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  "include": ["src/**/*", "types/**/*"],
+  "compilerOptions": {
+    "moduleResolution": "node",
+    "declaration": true,
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true,
+    "resolveJsonModule": true,
+    "strict": true,
+    "stripInternal": true,
+    "forceConsistentCasingInFileNames": true,
```
Comment (wkurniawan07) [inline tsconfig.base.json:11]: If you'd like to have more checks, e.g. the same level as our Angular 13 front-ends, you can consider adding these flags:
```json
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noImplicitReturns": true,
    "noFallthroughCasesInSwitch": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
```

**Pull request merged**